daemon/cmd/devices.go

```diff
@@ -0,0 +1,296 @@
+			for dev := range dm.devices {
+				option.Config.DirectRoutingDevice = dev
+				break
+			}
+		} else if k8sNodeLink != nil {
+			option.Config.DirectRoutingDevice = k8sNodeDev
+		} else {
+			return fmt.Errorf("Unable to determine direct routing device. Use --%s to specify them",
+				option.DirectRoutingDevice)
+		}
+		log.Infof("Detected %s=%s", option.DirectRoutingDevice, option.Config.DirectRoutingDevice)
+	}
+
+	if detectIPv6MCastDev {
+		if k8sNodeLink != nil && k8sNodeLink.Attrs().Flags&net.FlagMulticast != 0 {
+			option.Config.IPv6MCastDevice = k8sNodeDev
+			log.Infof("Detected %s=%s", option.IPv6MCastDevice, option.Config.IPv6MCastDevice)
+		} else {
+			return fmt.Errorf("Unable to determine Multicast device. Use --%s to specify them",
+				option.IPv6MCastDevice)
+		}
+	}
+
+	// TODO(JM): Remove mutation of option.Config.Devices as it is unsafe. A follow up will add support
+	// for runtime detection of new devices and will switch cilium agent to use GetDevices() instead.
+	option.Config.Devices = dm.getDevices()
+	log.WithField(logfields.Devices, option.Config.Devices).Info("Detected devices")
+
+	return nil
+}
+
+// GetDevices returns the current list of devices Cilium should attach programs to.
+func (dm *DeviceManager) GetDevices() []string {
+	dm.mu.Lock()
+	defer dm.mu.Unlock()
+	return dm.getDevices()
+}
+
+func (dm *DeviceManager) getDevices() []string {
+	devs := make([]string, 0, len(dm.devices))
+	for dev := range dm.devices {
+		devs = append(devs, dev)
+	}
+	sort.Strings(devs)
+	return devs
+}
+
+// isViableDevice returns true if the given link is usable and Cilium should attach
+// programs to it.
+func (dm *DeviceManager) isViableDevice(l3DevOK bool, link netlink.Link) bool {
+	name := link.Attrs().Name
+
+	// Do not consider any of the excluded devices.
+	for _, p := range excludedDevicePrefixes {
+		if strings.HasPrefix(name, p) {
+			log.WithField(logfields.Device, name).Debugf("Skipping device as it has excluded prefix '%s'", p)
+			return false
+		}
+	}
+
+	// Skip slave devices.
+	if link.Attrs().RawFlags&unix.IFF_SLAVE != 0 {
+		log.WithField(logfields.Device, name).Debugf("Skipping slave device")
+		return false
+	}
+
+	// Ignore L3 devices if we cannot support them.
+	if !l3DevOK && !mac.LinkHasMacAddr(link) {
+		log.WithField(logfields.Device, name).
+			Warn("Ignoring L3 device; >= 5.8 kernel is required.")
+		return false
+	}
+
+	return true
+}
+
+// updateDevicesFromRoutes processes a batch of routes and updates the set of
+// devices. Returns true if devices changed.
+func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.Route) bool {
+	linkIndices := make(map[int]struct{})
+
+	// Collect all link indices mentioned in the route update batch
+	for _, route := range routes {
+		// Only consider devices that have global unicast routes,
+		// e.g. skip loopback, multicast and link local routes.
+		if route.Dst != nil && !route.Dst.IP.IsGlobalUnicast() {
+			continue
+		}
```

Comment:
Hmm I don't quite follow. We want to ignore all uninteresting routes here and not consider the links that don't have interesting routes.

Reply:
Yes, I'm not proposing to change that. But these four lines could be moved inside the second loop and it would achieve the same while avoiding the need for `linkIndices`, no?

Reply:
I'm still not following how that would work. We may have hundreds of routes from which we want to create a set of interesting links to look at. The second loop doesn't deal with routes anymore so this check doesn't work there. Also please check the final code as the second commit modified this.

Reply:
> Also please check the final code as the second commit modified this.

Ah, forgot that. What I had in mind won't work with `linkInfo.hasDefaultRoute` :+1: